rt: add Handle::block_on

davidpdrsn opened the pull request:
This is a continuation of https://github.com/tokio-rs/tokio/pull/3549 but with tests for the behavior described [here](https://github.com/tokio-rs/tokio/pull/3549#issuecomment-785382050) and part of the implementation.

Still missing network operations failing if runtime associated with the handle has been shutdown. Changes to make timers and filesystem operations fail are at least passing the tests.

This is my first contribution to Tokio but let me know if I'm on the right track here 😊

Refs: #3097
Fixes #2965, #3096
cc @stuhood, @Keruspe

Darksonn commented on tokio/src/fs/mod.rs:
```diff
@@ -108,6 +108,11 @@ where
     F: FnOnce() -> io::Result<T> + Send + 'static,
     T: Send + 'static,
 {
+    let handle = crate::runtime::Handle::current();
+    if handle.is_shutdown() {
+        panic!("Cannot run I/O operations on a runtime that has been shutdown");
+    }
+
```
We want to keep all instances of these panic messages the same. See #3361.

Though in this case it probably should be panicking anyway inside the `sys::run` call. If it doesn't, fix `spawn_blocking` and this should fix itself.

davidpdrsn replied:
Do you mean that these panics should use `CONTEXT_MISSING_ERROR` as the message or that there should be constants for each message in `tokio/src/util/error.rs`?

carllerche replied:
I don’t think this should panic. Hitting this is not due to a user bug. This probably should be an Err

carllerche replied:
I'm not 100% sure 😆

carllerche replied:
Ok, based on the Discord convo, FS ops will already error when the runtime shuts down (by returning `Err`). We should keep this behavior as is. We may want to look to see if the behavior is properly documented.

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -201,6 +201,22 @@ impl Handle {
         handle
     }
+
+    /// TODO: write docs if this is a good direction
+    pub fn block_on<F: Future>(&self, future: F) -> F::Output {
```
Just adding this comment so we don't forget to fix doc.

Darksonn commented on tokio/src/time/driver/entry.rs:
```diff
@@ -543,6 +547,10 @@ impl TimerEntry {
         mut self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Result<(), super::Error>> {
+        if self.driver_shutdown() {
+            panic!("Cannot run timers on a runtime that has been shutdown");
+        }
```
Also #3361 here.

Darksonn commented on tokio/src/runtime/tests/handle.rs:
```diff
@@ -0,0 +1,296 @@
+
+// ==== nesting
+
+// TODO(david): this panics, should it?
+// TODO(david): test both for current thread and multi thread
```
Yes, calling `block_on` inside `block_on` should panic.

davidpdrsn replied:
👍

Darksonn commented on tokio/src/runtime/tests/handle.rs:
```diff
@@ -0,0 +1,296 @@
+    let _ = handle.block_on(TcpStream::connect("127.0.0.1:0"));
+}
+
+// NOTE: currently fails
+#[test]
+#[ignore]
+#[should_panic(expected = "some message about not being to do io from dropped runtime")]
+fn current_thread_block_on_with_net_created_before_shutdown_with_dropped_runtime_panics() {
+    let rt = new_current_thread();
+    let _enter = rt.enter();
+    let handle = rt.block_on(async { Handle::current() });
+
+    let net_future = TcpStream::connect("127.0.0.1:0");
+
+    rt.shutdown_timeout(Duration::from_secs(1000));
+
+    let _ = handle.block_on(net_future);
```
Should this panic? It might just need to return an IO error.

davidpdrsn replied:
It shouldn't panic to be consistent with timers and file io?

Darksonn reviewed (commented):
Some comments here.

LucioFranco commented on tokio/src/runtime/blocking/pool.rs:
```diff
@@ -260,6 +260,10 @@ impl Spawner {
             .unwrap()
     }
+
+    pub(crate) fn is_shutdown(&self) -> bool {
```
I wonder if we should use the `Spawner` to track the shutdown rather than the blocking pools spawner?

davidpdrsn replied:
Perhaps 🤷 There is already [`is_shutdown` on `Core`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/thread_pool/worker.rs#L54) which is accessible through `Handle.spawner`. Is that what you have in mind?

I guess we could also add an `AtomicBool` to [`Shared`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/thread_pool/worker.rs#L70) that gets set [here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/thread_pool/worker.rs#L770).

LucioFranco commented on tokio/src/time/driver/entry.rs:
```diff
@@ -494,6 +494,10 @@ impl TimerEntry {
     }
 
+    fn driver_shutdown(&self) -> bool {
+        self.driver.lock().is_shutdown
```
Do we want to move this to some sort of `AtomicBool`?

davidpdrsn replied:
Yeah I was wondering if this lock could be bad. The `is_shutdown` is defined [here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/time/driver/mod.rs#L149) which is accessible through an [`Arc<Mutex<_>>` here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/time/driver/handle.rs#L9), therefore the lock.

I guess we could move [`Inner.is_shutdown`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/time/driver/mod.rs#L135) to [`Handle`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/time/driver/handle.rs#L7) and make it an `AtomicBool`. Is that what you have in mind?

carllerche commented:
Generally I dont think attempting an op on a shutdown runtime should panic. Instead Err is better. That said... sleep is an interesting case because it does not return Result.

davidpdrsn commented:
Summary from discussions on Discord:

- `Handle::block_on` should simply return the output of the future.
- Things that aren't runtime dependent shouldn't care about the runtime being shutdown. They should continue to run.
  - This includes `tokio::signal::ctrl_c` and channels.
- Functions such as `TcpStream::connect` which return a result today and require the runtime to work should return an error if the runtime gets shutdown. This includes stuff in `tokio::net`.
  - Currently they will fail if put onto one runtime, moved to a second, and the first is shutdown. We should reuse the same error for this case.
- Futures created by functions from `tokio::fs` have two possible cases
  - If the task has started before the runtime is shutdown it should continue (cannot cancel in-flight).
  - Any task that has not yet started should error when given to `Handle::block_on` as the `JoinHandle` should return an error when awaited.
  - Same goes for `Stdin`, `Stderr`, `Stdout` as they also run on a blocking threadpool.
- Timer futures should panic as they don't return a `Result` today and are dependent on the runtime.
  - There is some uncertainty about whether this is the current behavior so we should verify this with tests.

carllerche commented:
@davidpdrsn what's the status of this? I think it was blocked on shutting down the timer properly. Is that issue tracked somewhere?

davidpdrsn commented:
@carllerche I haven't had more time to work in it. Will spend some time on it this week.

davidpdrsn commented:
I've pushed a bunch more changes. I think all the requirements should have tests and be working.

Still missing docs but I'll work on that if people think the implementation looks sound.

davidpdrsn commented on tokio/tests/rt_handle_block_on.rs:
```diff
@@ -0,0 +1,491 @@
+        let err = Handle::current().block_on(listener.accept()).unwrap_err();
+
+        assert_eq!(err.kind(), std::io::ErrorKind::Other);
+        assert_eq!(err.get_ref().unwrap().to_string(), "reactor gone");
```
This error message was already there. I think it makes sense to change it to `A Tokio 1.x context was found, but it is being shutdown.` for consistency. Wondering if that would be breaking change though.

LucioFranco commented on tokio/src/io/driver/registration.rs:
```diff
@@ -54,7 +54,7 @@ unsafe impl Sync for Registration {}
 impl Registration {
     /// Registers the I/O resource with the default reactor, for a specific
-    /// `Interest`. `new_with_interest` should be used over `new` when you need
+    /// `Interest`. `new_with_interest` should be ucrate::util::error::RUNTIME_SHUTTING_DOWN_ERRORsed over `new` when you need
```
```suggestion
    /// `Interest`. `new_with_interest` should be crate::util::error::RUNTIME_SHUTTING_DOWN_ERRORsed over `new` when you need
```

davidpdrsn replied:
Seems I got vim'ed.

Fixed in f2e44f92d76e0

jonhoo commented on tokio/src/io/driver/registration.rs:
```diff
@@ -232,7 +239,7 @@ cfg_io_readiness! {
 
             crate::future::poll_fn(|cx| {
-                if self.handle.inner().is_none() {
+                if self.handle.inner().is_none() || self.handle.is_shutdown() {
```
This seems a little unfortunate since `self.handle` also calls `self.handle.inner()`, so we end up upgrading inner twice.

davidpdrsn replied:
Ah yeah good catch. Should be fixed in 33fd64ebe08ee.

jonhoo commented on tokio/src/runtime/handle.rs:
```diff
@@ -201,6 +201,36 @@ impl Handle {
+            .expect("failed to park thread")
+    }
+
+    cfg_io_driver! {
```
Should this `cfg_` just be inside the `fn shutdown` around the `if let`?

davidpdrsn replied:
I did try that at first but the macro takes [an `item`](https://github.com/EmbarkStudios/tokio/blob/rt-handle-block-on/tokio/src/macros/cfg.rs#L65) so can't easily be put into functions afaik. It also checks [quite a few features](https://github.com/EmbarkStudios/tokio/blob/rt-handle-block-on/tokio/src/macros/cfg.rs#L64-L80) so didn't feel right copy pasting that.

jonhoo commented on tokio/src/time/driver/entry.rs:
```diff
@@ -543,6 +543,10 @@ impl TimerEntry {
         mut self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Result<(), super::Error>> {
+        if self.driver.is_shutdown() {
```
Are we concerned about the additional `inner.upgrade()` inside `is_shutdown` on every timer poll here?

davidpdrsn replied:
The `inner.upgrade()` is only on the io driver. This call is in the time driver which ends here https://github.com/EmbarkStudios/tokio/blob/rt-handle-block-on/tokio/src/time/driver/mod.rs#L452.

carllerche commented:
We need to slow down here. Anything that touches I/O shutdown needs to factor in #3481 or it will be racy. I will try to provide a deeper review shortly.

carllerche commented:
Thanks for the detailed tests. I am going to spend more time with this PR and will comment asap.

carllerche commented:
Sigh, I'm sorry, it seems like this PR is exposing more bugs in shutdown :grimacing:

carllerche commented on tokio/src/io/driver/registration.rs:
```diff
@@ -232,7 +239,8 @@ cfg_io_readiness! {
             crate::future::poll_fn(|cx| {
-                if self.handle.inner().is_none() {
+                let inner = self.handle.inner();
+                if inner.is_none() || inner.filter(|i| i.is_shutdown()).is_some() {
```
I was surprised that you needed to add this logic to make the tests pass. I did some digging and it appears that it is needed because the multi-threaded runtimes parker holds a strong ref to the I/O driver, which means that wakers hold strong refs to the driver. This goes back to #3481.

carllerche reviewed (changes requested):
Thanks for doing this. I left some notes inline.

This is definitely a bit of a tangled problem that you are tackling. The issue is that there are known bugs with the shutdown process that you are hitting when trying to write tests. I love that you took the initiative to try to fix I/O shutdown as part of this PR. Unfortunately, the proposed solution will be racy when concurrently shutting down and registering a new resource. Issue #3481 includes a proposed solution that should avoid races.

Fixing I/O shutdown is out of scope of *this* PR, but you also wrote great tests that won't pass until the shutdown bugs are fixed. I see two paths forward.

a) We pause this PR until #3481 is fixed, then rebase and land this one.
b) We mark tests that cannot pass until #3481 is fixed with `#[ignore]` and reference the issue. Then we revert the `is_shutdown: AtomicBool` part of this PR and land it w/ known issues.

If I had to pick one, I would opt for `b` as we do need `Handle::block_on` and the shutdown bugs are unrelated. Thoughts?

jonhoo commented:
I think I agree that b is the right path forward — we don't expect this to introduce _additional_ broken behavior, and we believe (I think) that we'd still want the same public API once we fix #3481.

davidpdrsn commented:
Yeah I think option b makes sense. I have removed the `is_shutdown: AtomicBool` part and marked the tests that failed with `#[ignore]` and added a little note about why.

I will write some initial docs for `Handle::block_on` today.

Darksonn commented:
If this doesn't introduce additional broken behavior, I'm fine with option b as well.

davidpdrsn commented:
I've written some initial docs.

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -202,7 +202,105 @@ impl Handle {
+    /// complete, and yielding its resolved result. Any tasks or timers which
+    /// the future spawns internally will be executed on the runtime.
+    ///
+    /// The behavior for multi threaded vs current thread schedulers is the same
+    /// as [`Runtime::block_on`]. See the docs of [`Runtime::block_on`] for more
+    /// details.
```
Is this true? Will it drive the IO driver when its the only call to block_on on a current_thread runtime?

davidpdrsn replied:
I remember checking that some time ago. But now I'm not sure. I'll have a look and perhaps add a test for it.

davidpdrsn replied:
Ah wait sorry. I commented on the wrong thing.

> Is this true? Will it drive the IO driver when its the only call to block_on on a current_thread runtime?

I'm not sure about this. Don't know enough about the internals to say 🤷 Do you know?

Darksonn replied:
The `Handle::block_on` in Tokio 0.2 didn't do this, but I don't know if that has changed since then. To be clear, it is plausible that it could have changed as part of `Runtime::block_on` being changed from `&mut self` to `&self`. But I don't know.

@carllerche do you know?

carllerche replied:
`Handle::block_on` and `Runtime::block_on` are slightly different. `Handle::block_on` **will not** drive the runtime where as `Runtime::block_on` **may** drive the runtime. Currently, this only happens with the `current_thread` scheduler.

davidpdrsn replied:
Alright. Is this something you think is worth mentioning or should I just cut out this bit?

Darksonn replied:
Yes, we should mention this.

davidpdrsn replied:
I've been thinking about this but sure how to word it accurately.

Darksonn replied:
How about this?

> When this is used on a current_thread runtime, only the `Runtime::block_on` method can drive the IO and timer drivers, but the `Handle::block_on` method cannot drive them. This means that, when using this method on a current_thread runtime, anything that relies on IO or timers will not work unless there is another thread currently calling `Runtime::block_on` on the same runtime.

carllerche replied:
I am applying ^^ to the PR so we can merge :)

carllerche replied:
Actually, it seems like I can't push to your PR :'(

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -202,7 +202,105 @@ impl Handle {
+    ///
+    /// # If the runtime has been shutdown
+    ///
+    /// If the `Handle`'s associated `Runtime` has been shutdown, through
+    /// [`Runtime::shutdown_background`] or [`Runtime::shutdown_timeout`], and
+    /// `Handle::block_on` is used it might return an error or panic. The exact
+    /// behavior depends on the types of futures used.
```
A runtime can also be shut down by dropping it.

davidpdrsn replied:
Good point! Added note about that.

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -202,7 +202,105 @@ impl Handle {
+    ///
+    /// ## Runtime independent futures
+    ///
+    /// Runtime independent futures will run as normal. They are not affected by
+    /// shutdown. This includes, but is not limited to, channels, signals, and
+    /// basic futures that don't actually `await` anything.
```
Does it include signals? Assuming you refer to `tokio::signal`.

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -202,7 +202,105 @@ impl Handle {
+    /// shutdown. This includes, but is not limited to, channels, signals, and
+    /// basic futures that don't actually `await` anything.
+    ///
+    /// ## [`spawn_blocking`] futures
+    ///
+    /// Futures created with [`spawn_blocking`] will run if they were started
+    /// before the runtime was shutdown. If they were created after the runtime
+    /// was shutdown they will get cancelled and the [`JoinHandle`] will return
+    /// a [`JoinError`].
```
This doens't seem relevant to the `block_on` docs.

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -202,7 +202,105 @@ impl Handle {
+    /// was shutdown they will get cancelled and the [`JoinHandle`] will return
+    /// a [`JoinError`].
+    ///
+    /// ## File system futures
+    ///
+    /// File system futures created by something in [`tokio::fs`] behave
+    /// similarly to [`spawn_blocking`] futures. They will run if started before
+    /// shutdown but fail with an error if started after shutdown.
+    ///
+    /// ## I/O future
+    ///
+    /// I/O futures created by something in [`tokio::net`] will return an error
+    /// regardless if the runtime was shutdown before or after the future was
+    /// created.
+    ///
+    /// ## Timer futures
+    ///
+    /// Timer futures created by something in [`tokio::time`] will panic if the
+    /// runtime has been shutdown. This is because the function signatures don't
+    /// allow returning errors.
```
Maybe just remove this, replacing it with a blanket statement that things that require a Tokio runtime will fail if used during shutdown of the runtime.

davidpdrsn replied:
I personally consider these important details and something that should be explained in the docs.

Darksonn replied:
Still, I would prefer saying something simpler like:

> If the runtime is shut down while a call to `block_on` is running, IO resources will return an error and timers will panic.

In particular, I don't think we should be explaining _why_ timers panic rather than return an error here.

davidpdrsn replied:
I think thats fine. Still gets the point across. I changed things in e1b6ebd3ce44e.

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -202,7 +202,105 @@ impl Handle {
+    ///
+    /// # Panics
+    ///
+    /// This function panics if the provided future panics, if called within an
+    /// asynchronous execution context, or if a timer future is executed on a
+    /// runtime that has been shutdown.
```
This could be simplified to "this function panics if the provided future panics". We don't have to list a few examples of why it might panic.

davidpdrsn replied:
The first half was copied from `Runtime::block_on` so if its relevant there I would consider it relevant here as well. In general I prefer when docs are specific about when things panic as it helps users know what to expect.

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -202,7 +202,105 @@ impl Handle {
+    /// // Create the runtime
+    /// let rt  = Runtime::new().unwrap();
+    ///
+    /// // Enter the runtime context. This is necessary for `Handle::curent` to
+    /// // work
+    /// let _enter = rt.enter();
+    ///
+    /// // Get a handle to the current runtime and execute the future, blocking
+    /// // the current thread until completion
+    /// Handle::current().block_on(async {
+    ///     println!("hello");
+    /// });
```
Maybe this example would be better as an example that uses `#[tokio::main]`?

davidpdrsn replied:
Sure 👍

carllerche commented on tokio/src/runtime/handle.rs:
```diff
@@ -201,6 +201,86 @@ impl Handle {
+        let _rt_enter = self.enter();
+
+        // Enter a **blocking** context. This prevents blocking from a runtime.
+        let mut _blocking_enter = crate::runtime::enter(true);
```
Nit: this does not need to be `_` prefixed (as it is used below).

davidpdrsn replied:
Oh yeah of course. Fixed!

carllerche reviewed (commented):
LGTM :+1: There is the doc tweak + a minor style nit, but everything else looks good to me. I will let @Darksonn take it from here.

carllerche commented:
It does look like CI needs to be fixed as well.

jeromegn commented:
FWIW, we've been running this in production on all our servers for 2 days.

carllerche commented on tokio/src/runtime/handle.rs:
```diff
@@ -202,6 +202,88 @@ impl Handle {
+    /// complete, and yielding its resolved result. Any tasks or timers which
+    /// the future spawns internally will be executed on the runtime.
+    ///
+    /// The behavior for multi threaded vs current thread schedulers is the same
+    /// as [`Runtime::block_on`]. See the docs of [`Runtime::block_on`] for more
+    /// details.
```
```suggestion
    /// When this is used on a `current_thread` runtime, only the
    /// [`Runtime::block_on`] method can drive the IO and timer drivers, but the
    /// `Handle::block_on` method cannot drive them. This means that, when using
    /// this method on a current_thread runtime, anything that relies on IO or
    /// timers will not work unless there is another thread currently calling
    /// [`Runtime::block_on`] on the same runtime.
```

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -202,6 +202,91 @@ impl Handle {
+    ///
+    /// #[tokio::main]
+    /// async fn main () {
+    ///     // Get a handle to the current runtime and execute the future, blocking
+    ///     // the current thread until completion
+    ///     Handle::current().block_on(async {
+    ///         println!("hello");
+    ///     });
```
This example would panic.
```suggestion
    ///     let handle = Handle::current();
    ///     std::thread::spawn(move || {
    ///         // Using Handle::block_on to run async code in the new thread.
    ///         handle.block_on(async {
    ///             println!("hello");
    ///         });
    ///     });
```
I think it would be good to not have `no_run` on the examples.
